maven-core/src/main/java/org/apache/maven/graph/DefaultGraphBuilder.java

```diff
@@ -361,76 +406,35 @@ private File getReactorDirectory( MavenExecutionRequest request )
         throws ProjectBuildingException
     {
         MavenExecutionRequest request = session.getRequest();
-
         request.getProjectBuildingRequest().setRepositorySession( session.getRepositorySession() );
 
-        List<MavenProject> projects = new ArrayList<>();
+        List<ProjectCollectionStrategy> projectCollectionStrategies = Arrays.asList(
+                projectlessCollectionStrategy, // 1. Collect project for invocation without a POM.
+                multiModuleCollectionStrategy, // 2. Collect projects for all modules in the multi-module project.
+                requestPomCollectionStrategy   // 3. Collect projects for explicitly requested POM.
+        );
 
-        // We have no POM file.
-        //
-        if ( request.getPom() == null )
+        for ( ProjectCollectionStrategy strategy : projectCollectionStrategies )
```

Comment:
This looks weird to me. Now all stragegies are executes, but I would expect that only one really applies.
Instead I would expect to see a swich/if-elseif codeblock selecting the right strategy per case.

Reply:
This logic has indeed changed, with the main reasoning being that it is possible that the second strategy [1] fails and that we will have to fallback on the third strategy [2]. The second strategy can fail when it finds that the multiModuleProjectDirectory is incorrect; aka [not linking to the actual root of the multi-module project](https://github.com/apache/maven/pull/373/files#diff-cdc90272b007188972c28fe364870fb1R73). It can also fail when a module is collected which [requires a plugin in the same multi-module](https://github.com/apache/maven/pull/373/files#diff-cdc90272b007188972c28fe364870fb1R133).

The for loop will break when it found the last strategy to be successful (meaning: one or more projects have been collected).

But maybe the for loop structuring does not improve readability, so we could rework it into an if-else structure as you suggest. WDYT @mthmulders ?

[1] The second strategy is most of what this PR is about: collecting projects based on the multiModuleProjectDirectory/pom.xml.
[2] The third strategy is collecting projects based on the pom in the request (basically the current 3.6.3 behavior).

Reply:
I thought I replied to this one already... Can't find it though.

In fact, the old code already had strategies, except they weren't implemented in separate classes, just all in this class. It wasn't very obvious from the old code, but in fact, it already employed two strategies: if there's no POM (say when you run `archetype:generate`) or when there is a POM. If the former would fail, the second would take over. Now that we're adding a third strategy, we felt we'd make this structure more explicit, including the fact that the strategies are executed one after each other.

Reply:
I don't mind moving the logic to separate classes, that'll probably improve readability and testability.
I really prefer to make the choice of the strategy explicit instead of magically breaking out of a for-loop.
It will also make it easier to identify the awkward usecase, making the discussion easier to drop this support in the future.

Reply:
@mthmulders and I discussed it offline, and we agree. Let's go for more readability and explictness. 
Can you please check the new commit whenever you have some time?